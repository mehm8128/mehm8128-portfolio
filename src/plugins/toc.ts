import type { Root, Element, ElementContent, Text } from "hast";
import { visit } from "unist-util-visit";

export const rehypeCollapsableToc = () => {
  return (tree: Root) => {
    const rootUlElement: Element = {
      type: "element",
      tagName: "ol",
      properties: {},
      children: [],
    };

    visit(tree, "element", (node) => {
      visitorCallback(node, rootUlElement);
    });

    const detailsElement = createCollapsableToc(rootUlElement);
    tree.children.unshift(detailsElement);
  };
};

export const visitorCallback = (node: Element, rootUlElement: Element) => {
  if (!/^h[2-6]$/.test(node.tagName)) {
    return;
  }

  const headingLevel = getHeadingLevelFromElement(node);
  const liElement = createListItemElement(node);
  const rootUlElementChildren = assertElementNodeList(rootUlElement.children);

  // h2ならolに直接追加
  if (headingLevel === 2) {
    rootUlElement.children.push(liElement);
    return;
  }

  // 一番新しいliから、同じレベルのheadingを入れているolを探す
  const sameLevelUlElement = searchSameLevelUlElement(
    rootUlElementChildren,
    headingLevel
  );
  if (sameLevelUlElement) {
    sameLevelUlElement.children.push(liElement);
    return;
  }

  // 一番新しいliの一番深いところに新しくolを作って追加
  const deepestLiElement = getDeepestLiElement(rootUlElementChildren);
  const newUlElement = createUlElement();
  newUlElement.children.push(liElement);
  deepestLiElement.children.push(newUlElement);
};

/**
 * 引数のolに入っている一番新しいliの中で、levelと同じ見出しレベルのli要素を返す
 */
const searchSameLevelUlElement = (
  rootUlElement: Element[],
  level: number
): Element | undefined => {
  const rootLiElement = assertElementNode(
    rootUlElement[rootUlElement.length - 1]
  );
  const headingAnchorElement = assertElementNode(rootLiElement.children[0]);
  const headingTextElement = assertElementText(
    headingAnchorElement.children[0]
  );
  const rootElementHeadingLevel = getHeadingLevelFromText(headingTextElement);
  if (rootElementHeadingLevel === level) {
    return rootLiElement;
  }

  if (rootLiElement.children[1] === undefined) {
    return;
  }
  const childUlElement = assertElementNode(rootLiElement.children[1]);

  return searchSameLevelUlElement(
    assertElementNodeList(childUlElement.children),
    level
  );
};

/**
 * 引数のolに入っている一番新しいliの中で、一番深いli要素を取得する
 */
const getDeepestLiElement = (rootUlElement: Element[]): Element => {
  const rootLiElement = assertElementNode(
    rootUlElement[rootUlElement.length - 1]
  );

  if (!rootLiElement.children[1]) {
    return rootLiElement;
  }
  const olElement = assertElementNode(rootLiElement.children[1]);

  return getDeepestLiElement(assertElementNodeList(olElement.children));
};

const getHeadingLevelFromElement = (headingElement: Element) => {
  const headingLevel = Number(headingElement.tagName.charAt(1));
  return headingLevel;
};
const getHeadingLevelFromText = (headingElement: Text) => {
  const headingLevel = Number(headingElement.value.charAt(1));
  return headingLevel;
};

const createUlElement = (): Element => {
  return {
    type: "element",
    tagName: "ol",
    properties: {},
    children: [],
  };
};

const createListItemElement = (node: Element): Element => {
  const headingChildren = node.children
    .map((child) => (child.type === "element" ? child.children : []))
    .flat();
  const headingTextElement = headingChildren.find(
    (child) => child.type === "text"
  );
  if (!headingTextElement) {
    throw new Error("見出しにテキストがありません");
  }
  const headingText = headingTextElement.value;

  const anchorElement: Element = {
    type: "element",
    tagName: "a",
    properties: { href: `#${headingText}` },
    children: [{ type: "text", value: headingText }],
  };

  return {
    type: "element",
    tagName: "li",
    properties: {},
    children: [anchorElement],
  };
};

const createCollapsableToc = (rootUlElement: Element): Element => {
  const summaryElement: Element = {
    type: "element",
    tagName: "summary",
    properties: {},
    children: [{ type: "text", value: "目次" }],
  };
  const detailsElement: Element = {
    type: "element",
    tagName: "details",
    properties: {},
    children: [summaryElement, rootUlElement],
  };

  return detailsElement;
};

const assertElementNode = (node: ElementContent): Element => {
  if (node.type !== "element") {
    throw new Error("Elementノードではありません");
  }
  return node;
};

const assertElementNodeList = (nodeList: ElementContent[]): Element[] => {
  if (nodeList.some((node) => node.type !== "element")) {
    throw new Error("Elementノードではありません");
  }
  return nodeList as Element[];
};

const assertElementText = (node: ElementContent): Text => {
  if (node.type !== "text") {
    throw new Error("Textノードではありません");
  }
  return node;
};
