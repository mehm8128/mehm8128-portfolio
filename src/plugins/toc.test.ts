import { describe, expect, test } from "vitest";
import { rehype } from "rehype";
import { rehypeCollapsableToc } from "./toc";

describe("rehypeCollapsableToc", () => {
  test("h2だけのとき、同じレベルで作成される", async () => {
    const input = `
      <h2 id="heading-1"><a href="#heading-1">Heading 1</a></h2>
      <h2 id="heading-2"><a href="#heading-2">Heading 2</a></h2>
      <h2 id="heading-3"><a href="#heading-3">Heading 3</a></h2>
      `;
    const { value } = await rehype().use(rehypeCollapsableToc).process(input);
    expect(value).toMatchSnapshot();
  });
  test("h2以外で同じレベルの見出しが適切に作成される", async () => {
    const input = `
      <h2 id="heading-1"><a href="#heading-1">Heading 1</a></h2>
      <h3 id="heading-1-1"><a href="#heading-1-1">Heading 1-1</a></h3>
      <h3 id="heading-1-2"><a href="#heading-1-2">Heading 1-2</a></h3>
      <h4 id="heading-1-2-1"><a href="#heading-1-2-1">Heading 1-2-1</a></h4>
      <h4 id="heading-1-2-2"><a href="#heading-1-2-2">Heading 1-2-2</a></h4>
    `;
    const { value } = await rehype().use(rehypeCollapsableToc).process(input);
    expect(value).toMatchSnapshot();
  });
  test("低いレベルの見出しに戻るときにも適切に作成される", async () => {
    const input = `
      <h2 id="heading-1"><a href="#heading-1">Heading 1</a></h2>
      <h3 id="heading-1-1"><a href="#heading-1-1">Heading 1-1</a></h3>
      <h4 id="heading-1-1-1"><a href="#heading-1-1-1">Heading 1-1-1</a></h4>
      <h3 id="heading-2"><a href="#heading-2">Heading 2</a></h3>
      <h4 id="heading-2-1"><a href="#heading-2-1">Heading 2-1</a></h4>
      <h2 id="heading-3"><a href="#heading-3">Heading 3</a></h2>
    `;
    const { value } = await rehype().use(rehypeCollapsableToc).process(input);
    expect(value).toMatchSnapshot();
  });
});
