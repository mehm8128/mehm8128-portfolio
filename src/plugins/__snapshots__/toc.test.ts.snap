// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rehypeCollapsableToc > h2だけのとき、同じレベルで作成される 1`] = `
"<details><summary>目次</summary><ol><li><a href="#Heading 1">Heading 1</a></li><li><a href="#Heading 2">Heading 2</a></li><li><a href="#Heading 3">Heading 3</a></li></ol></details><html><head></head><body><h2 id="heading-1"><a href="#heading-1">Heading 1</a></h2>
      <h2 id="heading-2"><a href="#heading-2">Heading 2</a></h2>
      <h2 id="heading-3"><a href="#heading-3">Heading 3</a></h2>
      </body></html>"
`;

exports[`rehypeCollapsableToc > h2以外で同じレベルの見出しが適切に作成される 1`] = `
"<details><summary>目次</summary><ol><li><a href="#Heading 1">Heading 1</a><ol><li><a href="#Heading 1-1">Heading 1-1</a><ol><li><a href="#Heading 1-2">Heading 1-2</a><ol><li><a href="#Heading 1-2-1">Heading 1-2-1</a><ol><li><a href="#Heading 1-2-2">Heading 1-2-2</a></li></ol></li></ol></li></ol></li></ol></li></ol></details><html><head></head><body><h2 id="heading-1"><a href="#heading-1">Heading 1</a></h2>
      <h3 id="heading-1-1"><a href="#heading-1-1">Heading 1-1</a></h3>
      <h3 id="heading-1-2"><a href="#heading-1-2">Heading 1-2</a></h3>
      <h4 id="heading-1-2-1"><a href="#heading-1-2-1">Heading 1-2-1</a></h4>
      <h4 id="heading-1-2-2"><a href="#heading-1-2-2">Heading 1-2-2</a></h4>
    </body></html>"
`;

exports[`rehypeCollapsableToc > 低いレベルの見出しに戻るときにも適切に作成される 1`] = `
"<details><summary>目次</summary><ol><li><a href="#Heading 1">Heading 1</a><ol><li><a href="#Heading 1-1">Heading 1-1</a><ol><li><a href="#Heading 1-1-1">Heading 1-1-1</a><ol><li><a href="#Heading 2">Heading 2</a><ol><li><a href="#Heading 2-1">Heading 2-1</a></li></ol></li></ol></li></ol></li></ol></li><li><a href="#Heading 3">Heading 3</a></li></ol></details><html><head></head><body><h2 id="heading-1"><a href="#heading-1">Heading 1</a></h2>
      <h3 id="heading-1-1"><a href="#heading-1-1">Heading 1-1</a></h3>
      <h4 id="heading-1-1-1"><a href="#heading-1-1-1">Heading 1-1-1</a></h4>
      <h3 id="heading-2"><a href="#heading-2">Heading 2</a></h3>
      <h4 id="heading-2-1"><a href="#heading-2-1">Heading 2-1</a></h4>
      <h2 id="heading-3"><a href="#heading-3">Heading 3</a></h2>
    </body></html>"
`;
